KAFKA-6455: Improve DSL operator timestamp semantics

mjsax opened the pull request:
Basic idea:
 - KTable-KTable join: set `max(left-ts,right-ts)` for result
 - #agg(...) (stream/table windowed/non-windowed): set `max(ts1, ts2, ts3,...)` of all input records that contribute to the aggregation result
 - for all stateless transformation: input-ts -> output-ts

mjsax committed:
KAFKA-6455: Improve DSL operator timestamp semantics

mjsax commented:
Call for review @guozhangwang @bbejeck @vvcephei @ableegoldman @cadonna @abbccdda

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableAggregate.java:
```diff
@@ -85,10 +85,19 @@ public void process(final K key, final Change<V> value) {
             if (value.oldValue != null && oldAgg != null) {
                 intermediateAgg = remove.apply(key, value.oldValue, oldAgg);
+                // TODO: not sure if this make sense
+                // technically, if we remove a value from an aggregation,
```
Yeah, I see your point here.

But I think this still makes sense from the standpoint as of now we never move back in time. We always set the timestamp to ` max(context.timestamp(), timestamp we've seen so far)` so this seems consistent with our current behavior.  In the end, I'm not entirely sure either.  

Additionally, we make the identical call on line 116 below, could we just consolidate and make the call ` newTimestamp = Math.max(context().timestamp(), oldAggAndTimestamp.timestamp());` once?   In practice would we ever hit the condition on line 91 and not line 116?

mjsax replied:
In current code, I think you are right -- however, I think it's better to keep both separate, as we might want to do an optimization to only send one `Change` record instead of two if the key for both is the same. Hence, I think the current code is more robust.

vvcephei replied:
Interesting!

Let's say that we're using this operator to count foreign key references. E.g.:
```
A -> C
B -> C
```
We have two records `A` and `B` referencing `C`. We'd group by the reference and then count, winding up with: `C:2` as the count.

Let's add some timestamps:
```
observe A -> C (ts 0)
result C:1 (ts 0)
observe B -> C (ts 1)
result C:2 (ts 1)
```
And then, to get to the point of the TODO, let's say that `B` gets deleted at time `2`. We need to subtract its contribution from the result, but we actually have 3 options for what to do with the timestamp:
* revert the result timestamp back to `0`
* leave the result timestamp alone at `1`. This option doesn't make much sense at all, I just mention it for completeness. Let's just forget about it now.
* update the result timestamp to the delete event's time, `2`

Whether you revert to `0` (what the TODO wants) or advance to `2` (what the code does) depends on what you think the result timestamp _means_. Reverting to `0` implies that the result timestamp captures the high-watermark time of all the updates that are currently contributing to the result. Advancing to `2` implies that the result timestamp is just the last time the result was updated.

Is there a standard SP definition of aggregation result processing timestamp to look to?

In the absence of one, we can just choose the semantics that we find most suitable. And actually, I kind of favor the latter. It's practical to compute (a valid consideration when you're designing a scalable SP system), and it's also a simpler definition, which improves the ergonomics of Streams. As a user, if I want to understand the result timestamp, I don't have to look back and find the highest timestamp of all the records that haven't subsequently been removed from the aggregation, I just have to look at the highest timestamp of any record that contributed to the aggregation at all.

And, of course, as Bill pointed out, time for a single key should always move forward or stay the same. If we roll the time back to `0`, then we're actually producing an out-of-order result (one with a later result but an earlier timestamp). If a reader of the data wants to enforce in-order updates (as it probably should), then it should discard the final `C:1 (ts: 0)` update, since it previously saw `C:2 (ts: 1)`, but this results in an incorrect final state of `C:2`.

mjsax replied:
> Is there a standard SP definition of aggregation result processing timestamp to look to?

Not really. The only definitions I am aware of, don't handle the delete case (most systems only support "fact streams" but no update semantics)

I guess that makes sense. I just wanted to have a discussion, because it's not completely straight forward. If we are all happy with taking "max" and to advance the timestamp, I have no objection at all.

vvcephei replied:
Sweet. Glad to have your expertise on this stuff!

big-andy-coates replied:
Max makes sense for me.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableReduce.java:
```diff
@@ -80,10 +80,19 @@ public void process(final K key, final Change<V> value) {
             // first try to remove the old value
             if (value.oldValue != null && oldAgg != null) {
                 intermediateAgg = removeReducer.apply(oldAgg, value.oldValue);
+                // TODO: not sure if this make sense
```
Same comments from above

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableTransformValues.java:
```diff
@@ -134,13 +135,17 @@ public void close() {
+            final ValueAndTimestamp<V> valueAndTimestamp = parentGetter.get(key);
+            return ValueAndTimestamp.make(
+                valueTransformer.transform(key, getValueOrNull(valueAndTimestamp)),
+                valueAndTimestamp == null ? context.timestamp() : valueAndTimestamp.timestamp());
```
If `valueAndTimestamp != null` should this be `max(valueAndTimestamp.timestamp(), context.timestamp())`

mjsax replied:
I don't think so. Note, this is a lookup into a store of a different upstream processor (in case, the KTable is not materialized). Thus, we should only return the timestamp that is in the store. It's not an output record we handle here, but an input record that we look up.

vvcephei replied:
Agreed.

On the opposite side of the boolean, if the `valueAndTimestamp` _is_ null, should we just return `null`?

mjsax replied:
> Agreed.

With whom?

> On the opposite side of the boolean, if the valueAndTimestamp is null, should we just return null?

 Why? If it's `null`, we should still pipe it through `transform` IMHO, and return whatever `transform()` returns.

vvcephei replied:
> Agreed.

Sorry, with you :)

On the `null` question: I see. There are some sketchy semantics around this, but that's not relevant to timestamps, but table/transform semantics in general, so we don't need to worry about it here.

I actually failed to remember that `make` returns `null` if the value is `null`, so this code already does what I was really concerned about. I thought it would return a `ValueAndTimestamp(null, context.timestamp())` if it got a null input, which seemed wrong. Now I see it will just return `null`, which I think is right.

So, we actually only return `context.timestamp()` if the transformer returns a non-null result for a null input. It seems a little strange to me, since `context.timestamp()` is actually unrelated to `key`'s time domain, but it probably doesn't matter, since:
* even with this change, we still are not actually _defining_ and time behavior, so we can do literally anything we want
* this method is currently only used for joins, where the resulting time would go on to be maxed with `context.timestamp()` anyway, resulting in `max(context.timestamp(), context.timestamp())`, which is trivially correct.

If anything, I'd weakly prefer to use `UNKNOWN` instead, but upon reflection, I don't think it matters much, if at all.

mjsax replied:
Ok. Updating this to `UNKNOWN`.

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/GlobalKTableJoinsTest.java:
```diff
@@ -41,76 +42,64 @@
-                results.put(key, value);
-            }
-        };
+        keyValueMapper = (key, value) -> value;
```
Thanks for the cleanup.

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamBranchTest.java:
```diff
@@ -46,24 +46,9 @@
-                return (key % 2) != 0;
-            }
-        };
+        final Predicate<Integer, String> isEven = (key, value) -> (key % 2) == 0;
```
Nice cleanup!

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:
```diff
@@ -118,7 +118,7 @@ public void testJoin() {
                 driver.pipeInput(recordFactory.create(topic1, expectedKeys[i], "A" + expectedKeys[i]));
             }
-            processor.checkAndClearProcessResult();
+            processor.checkAndClearProcessResult(new String[0]);
```
super nit: `new String[0]` used in several places here and below,  maybe create a `final` instance variable?

bbejeck reviewed (commented):
Thanks for this @mjsax, I've made an initial pass and just have a couple of comments, but overall this lgtm. I didn't get through all the test code, so I'll make another pass soon.

mjsax committed:
Github comments

mjsax commented:
Updated this

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableInnerJoin.java:
```diff
@@ -87,22 +91,27 @@ public void process(final K key, final Change<V1> change) {
+                newValue = joiner.apply(change.newValue, valueRight);
+                resultTimestamp = Math.max(context().timestamp(), valueAndTimestampRight.timestamp());
+            } else {
+                resultTimestamp = context().timestamp();
```
Should we still incorporate the right side's timestamp here? Otherwise, we might get time-travel.

```
observe RIGHT A:y (ts 2)
observe LEFT A:x (ts 0)
result A:(x,y) (ts 2)
observe LEFT A:null (ts 1)
result A:null (ts 2? or 1?)
```

Similar to the KTableAggregate, the result timestamp should probably be non-decreasing, so we probably want it to be `2`, not go backward to `1`.

Or does some other part of the join take care of this?

mjsax replied:
Well. If time is properly synchronized, the example you describe should never happen, because `LEFT a:null (1)` should have been processed before `RIGHT A:y (2)`. Of course, the scenario you describe is possible if we enforce processing after `max.idle.time.ms` passed.

Consider the right hand side timestamp is somewhat weird to me though, because the right hand side record has nothing to do with the result. It's an inner join, and the left hand side tombstone determines that the result is deleted at tombstone event time (at least that is my current reasoning), and thus the result tombstone should have the same timestamp as the input record tombstone. The out-of-order problem is not introduced in this operator, but the operator rather propagates the out-of-order issue downstream, what seems "correct" to me.

Happy to be convinced otherwise, however, I don't think that the join has to "fix" out-of-order for this case but should rather propagate it downstream.

big-andy-coates replied:
I agree with John that we don't want the ts of the result to go backwards.

> Consider the right hand side timestamp is somewhat weird to me though, because the right hand side record has nothing to do with the result. 

I'm not familiar with this code, but conceptually it seems to make sense that the result tombstone has a ts of max(last_result.ts, left.ts).

mjsax replied:
I am not convinced yet. Let's change the example a little bit and do:
```
observe RIGHT A:y (ts 20)
observe LEFT A:x (ts 0)
result A:(x,y) (ts 20)
observe LEFT A:null (ts 10)
result A:null (ts 10) // this is the tombstone in question -- alternative ts=20
observe RIGHT A:y2 (ts 15)
observe LEFT A:x2 (ts 17)
result A:(x1,x2) (ts 17)
```

With no out-of-order records, processing would be:
 - `LEFT A:x (ts 0)`
 - `LEFT A:null (ts 10)` // left table empty
 - `RIGHT A:y2 (ts 15)` -> result `A:(x,y2) (ts 15)`
 - `LEFT A:x2 (ts 17)` -> result `A:(x2,y2) (ts 17)`
 - `RIGHT A:y (ts 20)` -> result `A:(x2,y) (ts 20)`

For the processing with out-of-order data: If the result tombstone would be `A:null (ts 20)` it would semantically imply, that this tombstone deletes the final result  `A:(x1,x2) (ts 17)` (or actually correct result `A:(x2,y) (ts 20)`. In any case, the result is semantically empty. However, this sounds incorrect to me. The tombstone is rather a correction of the result table (that is at ts 20 already), but the correction _happens_ at event time 10. If there is no other update to the result KTable, the correction at ts=10 applies to the table at timestamp 20, too. However, if there is another update in between (as in our case, a new result at timestamp 17), the correction applies only from ts=10 to ts=17.

However, we can still see, that with out-of-order records, independent of the tombstone discussion, the compute result is not correct... However, fixing this is another mayor PR...

Bottom line: we are not able to handle out-of-order data gracefully for table-table-joins atm, and this PR won't fix it. We need to figure this out as follow up. However, I still think that using the input-tombstone-ts as output-tombston-ts is correct for this case.

Thoughts?

big-andy-coates replied:
Humm... couple of things...

First, if we take a subset of your example:

```
observe RIGHT A:y (ts 20)
observe LEFT A:x (ts 0)
result A:(x,y) (ts 20)
observe LEFT A:null (ts 10)
result A:null (ts 10) // this is the tombstone in question -- alternative ts=20
```

If the system goes with ts of 10 for the result tombstone, would not downstream consumers be right to ignore this update as its an update to a table entry with a timestamp less that the current value, i.e. they'd keep `result A:(x,y) (ts 20)` and ignore `result A:null (ts 10)` as its older.  Therefore their end state would be a table with `result A:null (ts 10)`.

A ts of 20 on the result tombstone would ensure the correct end state, as we'd fall back to offset order if timestamps match.

Secondly, a minor thing, but I *think* your 'with no out-of-order records' explanation incorrectly states the first RIGHT event would create a result, but I would of thought it would not as the LEFT is empty.  I'm not trying to nit-pick, just ensuring I understand right.

My corrected 'no out-of-order records', processing would be:

```
LEFT A:x (ts 0)
LEFT A:null (ts 10) // left table empty
RIGHT A:y2 (ts 15) // no output as left table empty
LEFT A:x2 (ts 17) -> result A:(x2,y2) (ts 17)
RIGHT A:y (ts 20) -> result A:(x2,y) (ts 20)
```

Thirdly, looking at your full example the second update to the right is `RIGHT A:y2 (ts 15)`, which has a lower timestamp to the first update.  For my own understanding, is this possible with KTables? I thought the table itself would ensure such out-of-order output from the table was not possible. I appreciate streams of data coming in to Kafka Streams often has out of data messages, but these messages we're talking about here are outputs from a table aggregation, right? (Forgive my ignorance if I'm way of the mark here).

Assuming the source tables are ensuring the change events coming out of them are only moving forward in time, then I believe having the result tombstone emitted with a ts of 20 would work, as the later joins would also produce results with ts of 20, and these would be retrained as they have a higher offset to the tombstone.

For example, If we take your example and run it through with ts 20 and correct ordering of RIGHT updates:

```
observe RIGHT A:y (ts 20)
observe LEFT A:x (ts 0)
result A:(x,y) (ts 20)
observe LEFT A:null (ts 10)
result A:null (ts 20) // switched to ts 20
                                  // no RIGHT update with ts < first.
observe LEFT A:x2 (ts 17)
result A:(x2,y) (ts 20) // wins over result tombstone with same ts due to higher offset.
```

I'm guessing my assumptions about the guarantees tables give on the ordering of their output events is wrong :(

mjsax replied:
Ack, my example was incorrect.... `RIGHT A:y2 (ts 15) // no output as left table empty` is correct.

>  I thought the table itself would ensure such out-of-order output from the table was not possible.

Out-of-order updates are indeed possible. Before KIP-258, updates to KTable (eg, reading from an input topic) are applied in offset order. Because timestamp information was not stored, it was not even possible to detect out-of-order updates and to discard them. With KIP-258, we had the discussion to now discard out-of-order updates of source-KTables, however, we decided to _not_ do this, due to backward compatibility concerns. We need to revisit this as a follow up...

About your first argument. I see your point. However, because the downstream KTable would apply the update (independent of the timestamp) as-of now, the incorrect result would be deleted. I agree, that we need to revisit this behavior, but it's out-of-scope for now.

To really fix the issue, I believe we actually need to distinguish between a "delete" and a "revert"...

big-andy-coates replied:
Seems to me that out-of-order messages on a single key should not be accepted on a table.  On a stream of data it is probably more common to have out-of-order updates than not, due to the distributed nature of modern systems. But we should require more strict ordering for a topic that represents a table.

mjsax replied:
While I agree, the problem is that the current behavior is to accept out-of-order updates to `KTables` when reading from a topic. This PR does not change this for backward compatibility reasons (we only log a WARN in `KTableSource`). IMHO, we need a KIP and an opt-in model to make this change (and I agree, that we should do this).

Hence, we somehow need to handle the out-of-order case in this PR... I am not 100% sure if the current code, or the proposal is the better approach to be honest. On the other hand, we don't have a public contract about timestamp propagation yet, and we might need to revisit this again later. Thus, whatever we decide, the impact should be minor.

While I see the point about "not going back in time" and I also see the argument about 

> If the system goes with ts of 10 for the result tombstone, would not downstream consumers be right to ignore this update as its an update to a table entry with a timestamp less that the current value, i.e. they'd keep result A:(x,y) (ts 20) and ignore result A:null (ts 10) as its older. Therefore their end state would be a table with result A:null (ts 10).

I still lean slightly to the current behavior of the PR. But if you insist to change it, I am also ok, as I think we need to change it again later on anyway...

vvcephei replied:
Woah, awesome discussion.

Yeah, the "not going back in time" thing for KTables is not currently implemented, but it does seem like it's a precondition to having well-defined time semantics. So, we should consider join semantics assuming "correct" inputs from the source KTables. I.e., we should assume no out-of-order updates on the inputs to this join. In that case, then Andy's example still works , and we still get the correct final output.

The "incorrect" result from Matthias's example is an artifact of the "incorrect" KTable semantics. It doesn't seem like we should be "sloppy" with joins just because we already decided to be "sloppy" with Tables. There's no easy solution to the table semantics, but there is for the joins.

Another way to look at it is, if we "fix" the join time semantics, then all we have to do to get a sound system is to fix the table semantics. But if we leave the join semantics as is in this PR, then we have to fix _both_ tables and joins later on.

Does that seem to add up?

mjsax replied:
It does make sense.

One thing to note is, that we use the input tombstone ts for output ts in other joins, too. I think we should at least be consistent. On the other joins, nobody raised concerns about using the tombstone timestamp, hence, it seem to be fine.

Might also be good if you look into the corresponding tests, and comment on the line for which you think a different timestamp (and which one) would be better?

bbejeck replied:
While I also feel that the ts should not go backwards since we'll follow up work to fix out-of-order data, I'd be ok with the PR as is.

big-andy-coates replied:
I'm logging off to do bedtime stories I'm afraid, so don't have time to annotate tests.

I'm with John when it comes to making this as correct and as inline with how we want the system to work long term, if we can.

Though I can see benefit in consistency with other join types, I think that being consistently wrong is probably worse than being inconsistent. ;)

I think a good litmus is whether someone can come along and materialize the output into some traditional database using the key and the timestamp and end up with the right result.  This is likely a fairly common thing people want to do, so should work as expected.

I know this has got to be merged asap to meet deadlines, so its probably worth mentioning that I'm not strongly opposed to merged things as they are, as  there is an understanding that these semantics are not right and will need addressing.

mjsax replied:
Well. It's easy to change in all joins... all part of this PR. Let's see what @bbejeck thinks. I am 50:50 and it seems that @vvcephei and @big-andy-coates have strong opinion and good arguments to change it. I am inclined to update this atm.

\cc @guozhangwang If you want to chime in.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableOuterJoin.java:
```diff
@@ -84,12 +89,28 @@ public void process(final K key, final Change<V1> change) {
+
+            final long timestamp1;
+            if (change.newValue == null) {
+                timestamp1 = UNKNOWN;
```
Not sure if this is right... Even if the current value on the left is `null`, the update has come from somewhere, and I think it must have been triggered from the left, which means we actually got a `null` record with a known timestamp. Should we use this timestamp instead of `UNKNOWN` as the join result?

It seems weird to produce `-1` timestamps, since we can't handle them.

mjsax replied:
Hmmm... That's a fair point. I don't think that we would ever propagate `-1` downstream (at least we don't see any in the tests... of course, tests might miss a case). However, the question is, what the result timestamp for the output should be. In the current PR, the result timestamp would be the timestamp from the right hand side (because it's an outer join -- `max(-1, RHS-ts)`). It might make sense to change it to the tombstone timestamp though. Thoughts?

vvcephei replied:
SGTM. It might be that we don't propagate it anyway, but it looks like you changed it to use the tombstone's timestamp, which seems right to me.

mjsax replied:
Yes, I did update it. As I mentioned above (and thinking about it more), I agree that using the tombstone timestamp is better.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableOuterJoin.java:
```diff
@@ -126,16 +147,16 @@ public void init(final ProcessorContext context) {
             }
 
-            return newValue;
+            return ValueAndTimestamp.make(newValue, -1L);
```
Interesting. Why is this different than the other joins, where we look up both getter timestamps and compute the max?

mjsax replied:
Ups. This seems wrong... Guess I put a surrogate `-1` to make it compile at some point but forget to fix it...

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableRightJoin.java:
```diff
@@ -84,20 +88,30 @@ public void process(final K key, final Change<V1> change) {
+            if (change.newValue != null) {
+                resultTimestamp = Math.max(context().timestamp(), valueAndTimestampLeft.timestamp());
+            } else {
+                resultTimestamp = valueAndTimestampLeft.timestamp();
```
Similar question here. Even if the left value is null, should the `null` record's timestamp still contribute to the join result time?

mjsax replied:
Note, that this processor implement the right-hand-side input for a left-join. Hence, `change.newValue` is an update to the right hand side. If it's `null`, we get `<key:left-null>` as result and the current reasoning is that this should use the timestamp from the left hand side, similar to outer join. However, following you other comment, I agree that it might make sense to actually use the tombstone timestamp for the result in this case. Thought?

vvcephei replied:
SGTM. Thanks! (and thanks for the reasoning as well)

mjsax replied:
Sure. As you might have noticed, I change this to use the tombstone timestamp (ie, `context.timestamp()`, now. Similar to https://github.com/apache/kafka/pull/6725#discussion_r284432739

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableMapValues.java:
```diff
@@ -81,6 +81,18 @@ private V1 computeValue(final K key, final V value) {
+            timestamp = valueAndTimestamp.timestamp();
+        }
+
+        return ValueAndTimestamp.make(newValue, timestamp);
```
This will always produce a non-null ValueAndTimestamp result, possibly with a `null` value and a `0` timestamp, if it wasn't resolved upstream... Is this right?

mjsax replied:
No. `ValueAndTimestamp` will never carry a `null` value. `make` would return `null` if `newValue` is null.

vvcephei replied:
Ah, my mistake!

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableRepartitionMap.java:
```diff
@@ -101,19 +104,24 @@ public void process(final K key, final Change<V> change) {
+            final ValueAndTimestamp<V> valueAndTimestamp = parentGetter.get(key);
+            return ValueAndTimestamp.make(
+                mapper.apply(key, getValueOrNull(valueAndTimestamp)),
+                valueAndTimestamp == null ? context.timestamp() : valueAndTimestamp.timestamp());
```
If there was no mapping for the key upstream, should we just return null?

mjsax replied:
Not sure if I can follow. If `valueAndTimestamp == null`, we effectively call ` mapper.apply(key, null))`. If `mapper` returns `null`, `make()` will return `null` (and it does not matter which timestamp we pass into `make` as it's ignored anyway). If `mapper` does not return `null`, we need to put a timestamp: if `valueAndTimestamp` has a timestamp we use it, otherwise we fall back to `context.timestamp()`.

Does this make sense?

vvcephei replied:
Yeah, this was my same mistake about the behavior of `make`. Sorry for the confusion.

mjsax replied:
No need to apologize. Reviewing is asking questions and raising concerns :)

vvcephei reviewed (commented):
Hey @mjsax , Thanks for doing this! I have some questions about the nuances of the time handling, but overall, this is awesome.

I didn't look at the test code yet, in case it changes as a result of these comments.

mjsax committed:
Github comments

mjsax committed:
Github comments

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableAggregateTest.java:
```diff
@@ -85,33 +85,41 @@ public void testAggBasic() {
 
         driver.setUp(builder, stateDir, Serdes.String(), Serdes.String());
 
+        driver.setTime(10L);
```
These will need updating if we merge https://github.com/apache/kafka/pull/6732 . Sorry :/

mjsax replied:
Can we maybe merge 6732 this one?

vvcephei reviewed (approved):
I still favor taking the max timestamp in the table-join (https://github.com/apache/kafka/pull/6725/files#r284412392), but my approval isn't conditioned on it. Do what you think is right ;)

Thanks, Matthias! This is exciting.

big-andy-coates commented:
My vote if for max timestamp too, but likewise my approval isn't conditional. (Please do me a favour and let me feel like I have the power of 'approval' here ;) ).

bbejeck reviewed (approved):
Thanks for this @mjsax!  Overall I still prefer max timestamps, but I won't insist on that for this to move forward.

mjsax committed:
Github comments

mjsax commented:
Thanks @vvcephei @big-andy-coates @bbejeck 

Updated to use "max" for input tombstones for all joins. Also changed the getter for KTable-TransformValues to UNKOWN.

Hope we can merge this after build is green.

mjsax committed:
Fix findbug

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableAggregate.java:
```diff
@@ -85,10 +85,12 @@ public void process(final K key, final Change<V> value) {
             final ValueAndTimestamp<T> oldAggAndTimestamp = store.get(key);
             final T oldAgg = getValueOrNull(oldAggAndTimestamp);
             final T intermediateAgg;
+            long newTimestamp = context().timestamp();
```
nit: to be consistent and also easier reading for others, we could do the same pattern like above classes, i.e. not initialize the value, just assign within the if-else.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableAggregate.java:
```diff
@@ -104,13 +106,16 @@ public void process(final K key, final Change<V> value) {
                 }
 
                 newAgg = add.apply(key, value.newValue, initializedAgg);
+                if (oldAggAndTimestamp != null) {
```
I found this logic is a bit hard to read, eventually it means `if (oldValue != null || newValue != null) we always check if oldAggAndTimestamp should be updated`.

Could we just extract this part independently for reading purposes?

mjsax replied:
Not sure if I can follow. Can you elaborate?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableReduce.java:
```diff
@@ -95,14 +97,15 @@ public void process(final K key, final Change<V> value) {
                     newAgg = value.newValue;
                 } else {
                     newAgg = addReducer.apply(intermediateAgg, value.newValue);
+                    newTimestamp = Math.max(context().timestamp(), oldAggAndTimestamp.timestamp());
```
Ditto here.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableSource.java:
```diff
@@ -101,8 +99,17 @@ public void process(final K key, final V value) {
+                if (oldValueAndTimestamp != null) {
+                    oldValue = oldValueAndTimestamp.value();
+                    if (context().timestamp() < oldValueAndTimestamp.timestamp()) {
+                        LOG.warn("Detected out-of-order KTable update for {} at offset {}, partition {}.",
```
I know I suggested this, but just clarifying: in the task's record queue, we already detect lateness like this and also record in the `record-lateness` metric already, so here we just log it for KTable source specifically in addition to the general metrics.

mjsax replied:
Well. We record `record-lateness` in `PartitionGroup`. Here we don't record any metric but only issue a WARN log. In general, out-of-order data are ok, but for this case it's a correctness concern, and hence the warning.

I would actually like to take this one step further (in a follow up KIP), to allow people to opt-in and drop those records.

guozhangwang reviewed (commented):
Made a pass on non-testing classes.

bbejeck commented:
Merged #6725 into trunk
